### PR TITLE
27 create account

### PR DESCRIPTION
**Pull request by tamari-gray**
fixes *27

### CODE REVIEW

**Comment by tamari-gray** [comment]
couldnt help myself,jumped straight in the code as soon as i woke up haha

File: test/auth_section/create_account/widget_test.dart [hunk 1]
```diff
@@ -2,16 +2,108 @@ import 'package:flutter/material.dart';
 import 'package:flutter_test/flutter_test.dart';
 import 'package:mockito/mockito.dart';
 import 'package:niira/screens/create_account.dart';
+import 'package:niira/services/auth/auth_service.dart';
+import 'package:niira/services/auth/navigation_service.dart';
+import 'package:provider/provider.dart';
+
+import '../sign_in/widget_test.dart';
 
 class MockNavigatorObserver extends Mock implements NavigatorObserver {}
```
Comment (nickmeinhold) [inline test/auth_section/create_account/widget_test.dart:10]: I don't think you use this do you?

File: test/auth_section/create_account/widget_test.dart [hunk 2]
```diff
@@ -2,16 +2,108 @@ import 'package:flutter/material.dart';
...
 
+class MockNavService extends Mock implements NavigationService {}
+
 void main() {
+  Widget makeTestableCreateAccountWidget(
+      NavigationService mockNav, MockFirebaseAuthService mockAuth) {
```
Comment (nickmeinhold) [inline test/auth_section/create_account/widget_test.dart]: The type is NavigationService but the variable name suggests it's a mock - not sure if the type is meant to be MockNavigationService or the variable is named incorrectly?

File: test/auth_section/create_account/widget_test.dart [hunk 3]
```diff
@@ -2,16 +2,108 @@ import 'package:flutter/material.dart';
...
+          Provider<AuthService>(create: (_) => mockAuth),
+        ],
+        child: MaterialApp(
+          navigatorKey: mockNav.navigatorKey,
+          home: CreateAccountScreen(),
+        ));
+  }
+
+  group('auth tests', () {
+    testWidgets('navigate to lobby on successfull login',
```
Comment (nickmeinhold) [inline test/auth_section/create_account/widget_test.dart]: is the test description as you meant it? this is testing account creation rather than a login right?

File: test/auth_section/create_account/widget_test.dart [hunk 4]
```diff
@@ -2,16 +2,108 @@ import 'package:flutter/material.dart';
...
+        (WidgetTester tester) async {
+      //set up for testing
+      final _mockNavService = MockNavService();
+      final _mockAuthService = MockFirebaseAuthService(_mockNavService, true);
```
Comment (nickmeinhold) [inline test/auth_section/create_account/widget_test.dart]: `MockFirebaseAuthService` here is actually a fake rather than a mock right?  Also I think it makes more sense to fake the AuthService rather than FirebaseAuthService, as the fake doesn't need to know anything about the implementation. I think it's worth naming fakes with 'fake' in the name rather than 'mock', it's simple but effective documentation IMO.

File: test/auth_section/create_account/widget_test.dart [hunk 5]
```diff
@@ -2,16 +2,108 @@ import 'package:flutter/material.dart';
...
+    testWidgets('auth error: shows dialog with error message',
+        (WidgetTester tester) async {
+      //set up for testing
+      final _mockNavService = MockNavService();
+      // return firebase auth error
+      final _mockAuthService = MockFirebaseAuthService(_mockNavService, false);
+      await tester.pumpWidget(
+          makeTestableCreateAccountWidget(_mockNavService, _mockAuthService));
+
+      // submit valid credentials
```
Comment (nickmeinhold) [inline test/auth_section/create_account/widget_test.dart]: these are actually invalid credentials I think? In order to get an error thrown?

File: test/auth_section/create_account/widget_test.dart [hunk 6]
```diff
@@ -2,16 +2,108 @@ import 'package:flutter/material.dart';
...
+import 'package:provider/provider.dart';
+
+import '../sign_in/widget_test.dart';
 
 class MockNavigatorObserver extends Mock implements NavigatorObserver {}
 
+class MockNavService extends Mock implements NavigationService {}
+
 void main() {
+  Widget makeTestableCreateAccountWidget(
```
Comment (nickmeinhold) [inline test/auth_section/create_account/widget_test.dart:15]: Nice code re-use!

File: test/auth_section/sign_in/widget_test.dart [hunk 7]
```diff
@@ -27,6 +27,28 @@ class MockFirebaseAuthService extends AuthService {
   @override
   Stream<UserData> get streamOfAuthState {}
 
+  @override
```
Comment (nickmeinhold) [inline test/auth_section/sign_in/widget_test.dart:21]: It's not really part of this PR but for clarity I think the `MockFirebaseAuthService` class should be `FakeAuthService` and `implement` rather than `extend` AuthService. You will get the same result but it will be clearer what the intention is, IMO.

File: test_driver/mocks/services/mock_auth_service.dart [hunk 8]
```diff
@@ -14,6 +14,9 @@ class MockAuthService implements AuthService {
   @override
```
Comment (nickmeinhold) [inline test_driver/mocks/services/mock_auth_service.dart:14]: I think this should be FakeAuthService rather than MockAuthService - in Mockito terms I think this is a "real object" rather than a Fake but I think calling it a fake is the clearest documentation, IMO.

**Review by nickmeinhold — DISMISSED** [review]
Awesome work mate!!! And so fast - make sure you get some rest soon hey :-) 

Lots of comments but everything was just clarity/doco type comments I think, no blockers from memory.  You've got some great momentum going!

**Comment by nickmeinhold** [comment]
@tamari-gray I will leave merging for you if that's cool so you can decide whether or not you want to make changes from the comments.  Great work mate!

**Comment by tamari-gray** [comment]
Wooh I got some work to do :) Thanks for the detailed comments!

File: test/auth_section/create_account/widget_test.dart (hunk 3 above)
Comment (tamari-gray) [reply test/auth_section/create_account/widget_test.dart]: good eyes 😄

File: test/auth_section/sign_in/widget_test.dart (hunk 7 above)
Comment (tamari-gray) [reply test/auth_section/sign_in/widget_test.dart:21]: just read the mockito fake documentation, thanks for the suggestion! its alot cleaner only rewriting the methods im using for the test.

**Comment by tamari-gray** [comment]
can i please get another review @nickmeinhold 🙏 :)

File: test/auth_section/create_account/widget_test.dart [hunk 9]
```diff
@@ -14,23 +13,23 @@ class MockNavService extends Mock implements NavigationService {}
 
 void main() {
   Widget makeTestableCreateAccountWidget(
-      NavigationService mockNav, MockFirebaseAuthService mockAuth) {
+      MockNavService mockNavService, FakeFirebaseAuthService mockAuth) {
```
Comment (nickmeinhold) [inline test/auth_section/create_account/widget_test.dart]: So in my mind `FakeFirebaseAuthService mockAuth` should be `FakeAuthService fakeAuth`, but: 
- please just ignore me if you prefer a different naming pattern 
- am I right in thinking we have an abstract AuthService and a FirebaseAuthService that implements the interface? I can't remember... if so, same comment as earlier - the fake doesn't need to know how the service is implemented  so it makes more sense to fake the AuthService IMO

Comment (nickmeinhold) [reply test/auth_section/create_account/widget_test.dart]: Actually given that you did `MockNavService mockNavService` I guess the other fake would be `FakeAuthService fakeAuthService`

Comment (nickmeinhold) [reply test/auth_section/create_account/widget_test.dart]: I get a bit hung up on naming, as I'm sure you can tell - I just think it's super important given how much time we spend trying to understand existing code. Please don't hesitate to ignore my suggestions

**Review by nickmeinhold — DISMISSED** [review]
Changes look good to me mate!

**Comment by nickmeinhold** [comment]
@tamari-gray leaving for you to merge again so you can decide if you want to make any changes.  Thanks!

File: test/auth_section/create_account/widget_test.dart (hunk 9 above)
Comment (tamari-gray) [reply test/auth_section/create_account/widget_test.dart]: Yes you're right! Idk why I chose to mock the firebaseauthservice over the abstract class.
